src/appleseed/renderer/modeling/bsdf/glossylayerbsdf.cpp

```diff
@@ -0,0 +1,513 @@
+        }
+
+        void attenuate_substrate(
+            const void*                 data,
+            const foundation::Basis3f&  shading_basis,
+            const foundation::Vector3f& outgoing,
+            const foundation::Vector3f& incoming,
+            Spectrum&                   value) const override
+        {
+            do_attenuate_substrate(data, shading_basis, outgoing, incoming, value);
+        }
+
+        void attenuate_substrate(
+            const void*                 data,
+            const foundation::Basis3f&  shading_basis,
+            const foundation::Vector3f& outgoing,
+            const foundation::Vector3f& incoming,
+            DirectShadingComponents&    value) const override
+        {
+            do_attenuate_substrate(data, shading_basis, outgoing, incoming, value);
+        }
+
+      private:
+        typedef GlossyLayerBSDFInputValues InputValues;
+
+        static void apply_energy_compensation_factor(
+            const InputValues*          values,
+            const Vector3f&             outgoing,
+            const Vector3f&             n,
+            Spectrum&                   value)
+        {
+            const float Ess = get_directional_albedo(
+                std::abs(dot(outgoing, n)),
+                values->m_roughness);
+
+            if (Ess == 0.0f)
+                return;
+
+            const float fms = values->m_precomputed.m_F0 * (1.0f - Ess) / Ess;
+            value *= 1.0f + fms;
+        }
+
+        template <typename SpectrumType>
+        void do_attenuate_substrate(
+            const void*                 data,
+            const foundation::Basis3f&  shading_basis,
+            const foundation::Vector3f& outgoing,
+            const foundation::Vector3f& incoming,
+            SpectrumType&               value) const
+        {
+            const InputValues* values = static_cast<const InputValues*>(data);
+
+            const Vector3f& normal = shading_basis.get_normal();
+            const float cos_on = dot(normal, outgoing);
+            const float cos_in = dot(normal, incoming);
+
+            const float eta = 1.0f / values->m_ior;
+            const float Eo = get_dielectric_layer_directional_albedo(
+                eta,
+                values->m_roughness,
+                abs(cos_on));
+            const float Ei = get_dielectric_layer_directional_albedo(
+                eta,
+                values->m_roughness,
+                abs(cos_in));
+
+            value *= min(1.0f - Ei, 1.0f - Eo);
+
+            if (values->m_thickness != 0.0f)
+            {
+                const float Lo = safe_rcp(abs(cos_on), 1e-6f);
+                const float Li = safe_rcp(abs(cos_in), 1e-6f);
+                float L = Li + Lo;
+
+                // For transmission, use the average length.
+                if (cos_in * cos_on <= 0.0f)
+                    L *= 0.5f;
+
+                value *= pow(values->m_transmittance, values->m_thickness * L);
```

Comment:
`std::pow()` (add `std::` prefix to all standard functions)

Reply:
Transmitance is a Spectrum here.